cni/pkg/nftables/host_sets.go

```diff
@@ -0,0 +1,385 @@
+
+// HostNftSetManager implements AddressSetManager using nftables sets
+type HostNftSetManager struct {
+	v4SetName   string
+	v6SetName   string
+	prefix      string
+	enableIPv6  bool
+	nftProvider NftProviderFunc
+}
+
+// NewHostSetManager creates a new nftables-based host set manager
+func NewHostSetManager(setPrefix string, enableIPv6 bool) (*HostNftSetManager, error) {
+	// Lets use the same naming convention as ipset
+	v4SetName := fmt.Sprintf("%s-v4", setPrefix)
+	v6SetName := fmt.Sprintf("%s-v6", setPrefix)
+
+	// Use the real implementation when the package level variable is not set
+	nftProvider := nftProviderVar
+	if nftProvider == nil {
+		nftProvider = func(family knftables.Family, table string) (builder.NftablesAPI, error) {
+			return builder.NewNftImpl(family, table)
+		}
+	}
+
+	manager := &HostNftSetManager{
+		v4SetName:   v4SetName,
+		v6SetName:   v6SetName,
+		prefix:      setPrefix,
+		enableIPv6:  enableIPv6,
+		nftProvider: nftProvider,
+	}
+
+	// Initialize the sets
+	if err := manager.initializeSets(); err != nil {
+		return nil, err
+	}
+
+	return manager, nil
+}
+
+// initializeSets creates the nftables table and sets
+func (h *HostNftSetManager) initializeSets() error {
+	nft, err := h.nftProvider(knftables.InetFamily, AmbientNatTable)
+	if err != nil {
+		return err
+	}
+
+	tx := nft.NewTransaction()
+
+	// Create table
+	tx.Add(&knftables.Table{})
+
+	// Create IPv4 set
+	tx.Add(&knftables.Set{
+		Name:    h.v4SetName,
+		Type:    "ipv4_addr",
+		Comment: knftables.PtrTo("UUID of the pods that are part of ambient mesh"),
+	})
+
+	// Create IPv6 set if enabled
+	if h.enableIPv6 {
+		tx.Add(&knftables.Set{
+			Name:    h.v6SetName,
+			Type:    "ipv6_addr",
+			Comment: knftables.PtrTo("UUID of the pods that are part of ambient mesh"),
+		})
+	}
+
+	if err := nft.Run(context.TODO(), tx); err != nil {
+		return fmt.Errorf("failed to initialize nftables sets: %w", err)
+	}
+
+	builder.LogNftRules(tx)
+	return nil
+}
+
+// AddIP adds an IP address to the appropriate set (v4 or v6)
+func (h *HostNftSetManager) AddIP(ip netip.Addr, ipProto uint8, comment string, replace bool) error {
+	ipToInsert := ip.Unmap()
+	setName := h.v4SetName
+	if ipToInsert.Is6() {
+		if !h.enableIPv6 {
+			log.Debugf("IPv6 is not enabled. Skipping the addition of IP: %v", ip)
+			return nil
+		}
+		setName = h.v6SetName
+	}
+
+	nft, err := h.nftProvider(knftables.InetFamily, AmbientNatTable)
+	if err != nil {
+		return err
+	}
+
+	tx := nft.NewTransaction()
+
+	// Add element to the set
+	tx.Add(&knftables.Element{
+		Set:     setName,
+		Key:     []string{ipToInsert.String()},
+		Comment: knftables.PtrTo(comment),
+	})
+
+	if err := nft.Run(context.TODO(), tx); err != nil {
+		return fmt.Errorf("failed to add IP %s to nftables set %s: %w", ip, setName, err)
+	}
+
+	builder.LogNftRules(tx)
+	return nil
+}
+
+// DeleteIP removes an IP address from the appropriate set
+func (h *HostNftSetManager) DeleteIP(ip netip.Addr, ipProto uint8) error {
```

Comment:
nit: Remove names for all these unused vars since it's just to fulfill an interface
```suggestion
func (h *HostNftSetManager) DeleteIP(ip netip.Addr, _ uint8) error {
```